compiler/rustc_middle/src/ty/consts.rs

```diff
@@ -298,7 +298,19 @@ impl<'tcx> Const<'tcx> {
             &hir::Path { res: Res::Def(DefKind::ConstParam, _), .. },
         )) = expr.kind
         {
-            span_bug!(expr.span, "try_from_lit: received const param which shouldn't be possible");
+            for (_, node) in tcx.hir().parent_iter(expr.hir_id) {
+                match node {
+                    // We need to allow this so that `const` generics can be used in default fields.
+                    hir::Node::Field(_) => break,
+                    hir::Node::Item(_) | hir::Node::Crate(_) => {
+                        span_bug!(
+                            expr.span,
+                            "try_from_lit: received const param which shouldn't be possible"
+                        );
+                    }
+                    _ => {}
+                }
+            }
```

Comment:
Before rebase (and creating the lint), `field: i32 = CONST_PARAM` was working fine, but after those two things I needed to do *something* for `from_anon_const` to not blow up in that case, which I'm confident we *should* support.